lib/Runtime/Base/CompactCounters.h

```diff
@@ -0,0 +1,269 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#pragma warning(push)
+#pragma warning(disable:6200) // C6200: Index is out of valid index range, compiler complains here we use variable length array
+
+namespace Js
+{
+    template<class T, typename CountT = T::CounterFields>
+    struct Counter
+    {
+
+        volatile uint8 fieldSize;
+        union {
+            uint8 u8fields[CountT::Max];
+            int8 i8fields[CountT::Max];
+            WriteBarrierPtr<uint16> u16Fileds;
+            WriteBarrierPtr<uint32> u32Fileds;
+            WriteBarrierPtr<int16> i16Fileds;
+            WriteBarrierPtr<int32> i32Fileds;
+        };
+
+
+        Counter()
+            :fieldSize(1)
+        {
+            memset(u8fields, 0, (uint8)CountT::Max);
+        }
+
+        void AllocCounters(T* host, uint8 newSize);
+
+        uint32 Get(CountT typeEnum) const
+        {
+            uint8 type = static_cast<uint8>(typeEnum);
+            uint8 localFieldSize = fieldSize;
+            uint32 value = 0;
+            if (localFieldSize == 1)
+            {
+                value = this->u8fields[type];
+            }
+            else if (localFieldSize == 2)
+            {
+                value = this->u16Fileds[type];
+            }
+            else
+            {
+                Assert(localFieldSize == 4);
+                value = this->u32Fileds[type];
+            }
+
+            // check if main thread has updated the structure
+            if (localFieldSize == fieldSize) 
+            {
+                return value;
+            }
+            else 
+            {
+                Assert(!ThreadContext::GetContextForCurrentThread());
+                AutoCriticalSection autocs(T::GetLock());
```

Comment:
Is there any interpreter perf impact from this lock?  Is it rare enough to be have to use a larger size that it doesn't matter?